### PR TITLE
Configure hibernate and add user entity

### PR DESCRIPTION
**Pull request by Cebix90**
Configure hibernate and add user entity with crud service
crud service tested

### CODE REVIEW

File: src/main/java/com/cebix/investmenttrackerapp/databaseutils/CustomUserDAO.java [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+            if(!userExists(customUser)) {
+                session = sessionFactory.openSession();
+                transaction = session.beginTransaction();
+                customUser.setPassword(passwordEncoder.encode(customUser.getPassword()));
+                session.merge(customUser);
+                transaction.commit();
+            } else {
+                throw new UserExistsException();
+            }
+        } catch (Exception e) {
```
Comment (blazejprzyluski) [inline src/main/java/com/cebix/investmenttrackerapp/databaseutils/CustomUserDAO.java:32]: Which method in the try block can result in an exception?

File: src/main/java/com/cebix/investmenttrackerapp/databaseutils/CustomUserDAO.java [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+            if (transaction != null) {
+                transaction.rollback();
+            }
+            throw e;
+        } finally {
+            if (session != null) {
+                session.close();
```
Comment (blazejprzyluski) [inline src/main/java/com/cebix/investmenttrackerapp/databaseutils/CustomUserDAO.java:39]: Could be used in try with resources

File: src/main/java/com/cebix/investmenttrackerapp/databaseutils/CustomUserDAO.java [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+        Session session = null;
+        try {
+            session = sessionFactory.openSession();
+            CriteriaBuilder criteriaBuilder = session.getCriteriaBuilder();
+            CriteriaQuery<CustomUser> userQuery = criteriaBuilder.createQuery(CustomUser.class);
+            Root<CustomUser> userRoot = userQuery.from(CustomUser.class);
+            userQuery.select(userRoot).where(criteriaBuilder.equal(userRoot.get("email"), email));
+            CustomUser customUser = session.createQuery(userQuery).getSingleResultOrNull();
+            return customUser;
+        } finally {
```
Comment (blazejprzyluski) [inline src/main/java/com/cebix/investmenttrackerapp/databaseutils/CustomUserDAO.java:54]: session could be used in try with resources

File: src/main/java/com/cebix/investmenttrackerapp/databaseutils/CustomUserDAO.java [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+        Transaction transaction = null;
+        try {
+            session = sessionFactory.openSession();
+            transaction = session.beginTransaction();
+            CustomUser customUserToDelete = findUserByEmail(customUserEmail);
+            if (customUserToDelete != null) {
+                session.remove(customUserToDelete);
+                transaction.commit();
+            }
+        } catch (Exception e) {
```
Comment (blazejprzyluski) [inline src/main/java/com/cebix/investmenttrackerapp/databaseutils/CustomUserDAO.java:72]: Which method in the try block can result in an exception?

File: src/main/java/com/cebix/investmenttrackerapp/databaseutils/CustomUserDAO.java [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+            if (transaction != null) {
+                transaction.rollback();
+            }
+            throw e;
+        } finally {
+            if (session != null) {
+                session.close();
```
Comment (blazejprzyluski) [inline src/main/java/com/cebix/investmenttrackerapp/databaseutils/CustomUserDAO.java:79]: could be closed in try with resources

File: src/main/java/com/cebix/investmenttrackerapp/databaseutils/CustomUserDAO.java [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+            }
+        }
+    }
+
+    public void updateUser(CustomUser customUser) {
```
Comment (blazejprzyluski) [inline src/main/java/com/cebix/investmenttrackerapp/databaseutils/CustomUserDAO.java:84]: why do you update entire customUser object?

File: src/main/java/com/cebix/investmenttrackerapp/databaseutils/CustomUserDAO.java [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+            transaction = session.beginTransaction();
+            CustomUser customUserToUpdate = findUserByEmail(customUser.getEmail());
+            if (customUserToUpdate != null) {
+                customUserToUpdate.setPassword(passwordEncoder.encode(customUser.getPassword()));
+                customUserToUpdate.setPortfolio(customUser.getPortfolio());
+                session.merge(customUserToUpdate);
+                transaction.commit();
+            }
+        } catch (Exception e) {
+            if (transaction != null) {
```
Comment (blazejprzyluski) [inline src/main/java/com/cebix/investmenttrackerapp/databaseutils/CustomUserDAO.java:98]: Which method in the try block can result in an exception?

File: src/main/java/com/cebix/investmenttrackerapp/databaseutils/CustomUserDAO.java [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+                transaction.rollback();
+            }
+            throw e;
+        } finally {
+            if (session != null) {
```
Comment (blazejprzyluski) [inline src/main/java/com/cebix/investmenttrackerapp/databaseutils/CustomUserDAO.java:103]: can be used in try with resources

File: src/main/java/com/cebix/investmenttrackerapp/databaseutils/CustomUserDAO.java [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+                session.close();
+            }
+        }
+    }
+
+    public CustomUser findUserByEmail(String email) {
```
Comment (blazejprzyluski) [inline src/main/java/com/cebix/investmenttrackerapp/databaseutils/CustomUserDAO.java:44]: no checking if user email is existing/non empty etc

File: src/main/java/com/cebix/investmenttrackerapp/exceptions/UserExistsException.java [hunk 10]
```diff
@@ -0,0 +1,8 @@
+package com.cebix.investmenttrackerapp.exceptions;
+
+public class UserExistsException extends RuntimeException {
+    @Override
+    public String getMessage() {
+        return "User with same email already exists. Please try with another email.";
```
Comment (blazejprzyluski) [inline src/main/java/com/cebix/investmenttrackerapp/exceptions/UserExistsException.java:6]: I would rephrase it to: Email already in use. Please try different email.

File: src/test/java/com/cebix/investmenttrackerapp/databaseutils/CustomUserDAOTests.java [hunk 11]
```diff
@@ -0,0 +1,98 @@
...
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.mockito.junit.jupiter.MockitoExtension;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.boot.test.context.SpringBootTest;
+import org.springframework.test.annotation.DirtiesContext;
+
+import static org.junit.jupiter.api.Assertions.*;
+
+@SpringBootTest
```
Comment (blazejprzyluski) [inline src/test/java/com/cebix/investmenttrackerapp/databaseutils/CustomUserDAOTests.java:14]: Why this annotation?

File: src/test/java/com/cebix/investmenttrackerapp/databaseutils/CustomUserDAOTests.java [hunk 12]
```diff
@@ -0,0 +1,98 @@
...
+@DirtiesContext(classMode = DirtiesContext.ClassMode.BEFORE_EACH_TEST_METHOD)
```
Comment (blazejprzyluski) [inline src/test/java/com/cebix/investmenttrackerapp/databaseutils/CustomUserDAOTests.java:15]: Why this annotation?

File: src/test/java/com/cebix/investmenttrackerapp/databaseutils/CustomUserDAOTests.java [hunk 13]
```diff
@@ -0,0 +1,98 @@
...
+@ExtendWith(MockitoExtension.class)
```
Comment (blazejprzyluski) [inline src/test/java/com/cebix/investmenttrackerapp/databaseutils/CustomUserDAOTests.java:16]: You aren't even using mocks in this test

**Review by blazejprzyluski — COMMENTED** [review]
Please write tests using DAO as if you would with regular database, next time we will talk I will show you test containers > this will create a database from docker image that will exist only for test purposes.

File: src/main/java/com/cebix/investmenttrackerapp/databaseutils/CustomUserDAO.java (hunk 1 above)
Comment (Cebix90) [reply src/main/java/com/cebix/investmenttrackerapp/databaseutils/CustomUserDAO.java:32]: no point in trying catch unknown exception, removed

File: src/main/java/com/cebix/investmenttrackerapp/exceptions/UserExistsException.java (hunk 10 above)
Comment (Cebix90) [reply src/main/java/com/cebix/investmenttrackerapp/exceptions/UserExistsException.java:6]: True, it looks nicer and more readable

File: src/main/java/com/cebix/investmenttrackerapp/databaseutils/CustomUserDAO.java (hunk 9 above)
Comment (Cebix90) [reply src/main/java/com/cebix/investmenttrackerapp/databaseutils/CustomUserDAO.java:44]: You're right. Added verification of input parameters before performing the operation

File: src/test/java/com/cebix/investmenttrackerapp/databaseutils/CustomUserDAOTests.java (hunk 13 above)
Comment (Cebix90) [reply src/test/java/com/cebix/investmenttrackerapp/databaseutils/CustomUserDAOTests.java:16]: Forgot to delete

File: src/test/java/com/cebix/investmenttrackerapp/databaseutils/CustomUserDAOTests.java (hunk 11 above)
Comment (Cebix90) [reply src/test/java/com/cebix/investmenttrackerapp/databaseutils/CustomUserDAOTests.java:14]: I thought this annotation in SpringBoot is necessary to integrate tests with database, but now I see tests work fine just by declaring newCustomerDAO

File: src/test/java/com/cebix/investmenttrackerapp/databaseutils/CustomUserDAOTests.java (hunk 12 above)
Comment (Cebix90) [reply src/test/java/com/cebix/investmenttrackerapp/databaseutils/CustomUserDAOTests.java:15]: Its unnecessary, to be deleted

File: src/main/java/com/cebix/investmenttrackerapp/databaseutils/CustomUserDAO.java (hunk 6 above)
Comment (Cebix90) [reply src/main/java/com/cebix/investmenttrackerapp/databaseutils/CustomUserDAO.java]: changed updateUser method to 3 separately methods which are responsibility for update each field in CustomUser

File: src/main/java/com/cebix/investmenttrackerapp/databaseutils/CustomUserDAO.java [hunk 14]
```diff
@@ -0,0 +1,123 @@
...
+            CustomUser customUser = session.createQuery(userQuery).getSingleResultOrNull();
+            return customUser;
+        }
+    }
+
+    public void deleteUser(String customUserEmail) {
+        try (Session session = sessionFactory.openSession()) {
+            Transaction transaction = session.beginTransaction();
+            CustomUser customUserToDelete = findUserByEmail(customUserEmail);
+            if (customUserToDelete != null) {
```
[the remark this answers is not in the document]
Comment (Cebix90) [reply src/main/java/com/cebix/investmenttrackerapp/databaseutils/CustomUserDAO.java:56]: Added else with throw new UserNotFoundException

**Review by blazejprzyluski — APPROVED** [review]
